pkg/deployments/scaler_test.go

```diff
@@ -89,3 +76,111 @@ func TestSetDesiredScale(t *testing.T) {
+// but once scaleDownDelay passes the scale down will be triggered.
+func TestNoScaleDownAfterScaleUp(t *testing.T) {
+	var lastScale int32
+	var scaleFuncCalled bool
+
+	var mockScaleMtx sync.Mutex // Mutex to protect shared state in mockScaleFunc
+	mockScaleFunc := func(n int32, atLeastOne bool) error {
+		mockScaleMtx.Lock()
+		defer mockScaleMtx.Unlock()
+		lastScale = n
+		scaleFuncCalled = true
+		return nil
+	}
+
+	// Create a new scaler
+	s := newScaler(2*time.Second, mockScaleFunc)
+
+	// Trigger a scale up
+	s.UpdateState(5, 1, 10)
+	s.SetDesiredScale(7)
+	time.Sleep(10 * time.Millisecond)
+
+	// Verify scale up worked as expected
+	mockScaleMtx.Lock() // Ensure consistency of the checked state
+	if scaleFuncCalled != true {
+		t.Errorf("expected scaleFuncCalled to be true, got %v", scaleFuncCalled)
+	}
+	if scaleFuncCalled && lastScale != 7 {
+		t.Errorf("expected lastScale to be 7, got %v", lastScale)
+	}
+	scaleFuncCalled = false
+	mockScaleMtx.Unlock()
+
+	// Trigger a scale down right after scale up and verify it won't happen
+	s.SetDesiredScale(3)
+	time.Sleep(1 * time.Second)
+	mockScaleMtx.Lock() // Ensure consistency of the checked state
+	if scaleFuncCalled != false {
+		t.Errorf("expected scaleFuncCalled to be false, got %v", scaleFuncCalled)
+	}
+	mockScaleMtx.Unlock()
+
+	// Trigger a scale down after 2 more seconds and verify scale down succeeded
+	time.Sleep(2 * time.Second)
+	s.SetDesiredScale(3)
+	mockScaleMtx.Lock() // Ensure consistency of the checked state
+	if scaleFuncCalled != true {
+		t.Errorf("expected scaleFuncCalled to be true, got %v", scaleFuncCalled)
+	}
+	mockScaleMtx.Unlock()
+}
+
+// Tests that scale down would still be delayed even if
+// current replicas has been stable for a duration beyond the scaleDownDelay
+func TestDelayedScaleDownAfterSustainedUsage(t *testing.T) {
+	var lastScale int32
+	var scaleFuncCalled bool
+
+	var mockScaleMtx sync.Mutex // Mutex to protect shared state in mockScaleFunc
+	mockScaleFunc := func(n int32, atLeastOne bool) error {
+		mockScaleMtx.Lock()
+		defer mockScaleMtx.Unlock()
+		lastScale = n
+		scaleFuncCalled = true
+		return nil
+	}
+
+	// Create a new scaler
+	s := newScaler(2*time.Second, mockScaleFunc)
+
+	// Trigger a scale up
+	s.UpdateState(5, 1, 10)
+	s.SetDesiredScale(7)
+
+	// Verify scale up worked as expected
+	mockScaleMtx.Lock() // Ensure consistency of the checked state
+	if scaleFuncCalled != true {
+		t.Errorf("expected scaleFuncCalled to be true, got %v", scaleFuncCalled)
+	}
+	if scaleFuncCalled && lastScale != 7 {
+		t.Errorf("expected lastScale to be 7, got %v", lastScale)
+	}
+	scaleFuncCalled = false
+	mockScaleMtx.Unlock()
+	s.UpdateState(7, 1, 10)
+	// Wait for scale down delay to pass
+	time.Sleep(2*time.Second + 10*time.Millisecond)
+
+	// Trigger a scale down
+	s.SetDesiredScale(3)
+	time.Sleep(1 * time.Second)
+	mockScaleMtx.Lock() // Ensure consistency of the checked state
+	if scaleFuncCalled != false {
```

Comment:
You could also use `require.True(t, scaleFuncCalled)` - https://pkg.go.dev/github.com/stretchr/testify/require#True